File: toolargetool/src/main/java/com/gu/toolargetool/Logger.java
package com.gu.toolargetool;

import android.util.Log;

/**
 * Created by sfriedenberg on 09/20/2017.
 */

public class Logger {
    protected int priority;
    protected String tag;

    public Logger(int priority, String tag) {
        this.priority = priority;
        this.tag = tag;
    }

    /**
     * override this method if you wish to use your own logging facilities
     *
     * @param msg
     */
    public void log(String msg) {
        Log.println(priority, tag, msg);
    }

    /**
     * In the case that the formatter throws an exception during formatting,
     * this method is called to log the exception. Like the above log method, you can override this too
     *
     * @param e
     */
    public void logException(Exception e) {
        Log.println(priority, tag, e.getMessage());
    }

    public static Logger DEFAULT_LOGGER = new Logger(Log.DEBUG, "TooLargeTool");
}

File: toolargetool/src/main/java/com/gu/toolargetool/Logger.kt
package com.gu.toolargetool

import android.util.Log

/**
 * Interface that allows flexibility in how TooLargeTool's output is logged. The default
 * implementation [LogcatLogger] should be suitable in most cases.
 */
interface Logger {
    fun log(msg: String)
    fun logTransactionTooLargeDetected(msg: String)
    fun logException(e: Exception)
}

/**
 * The default implementation of [Logger].
 *
 * @author [@sfriedenberg](https://github.com/friedenberg)
 */
class LogcatLogger(private val priority: Int = Log.DEBUG, private val tag: String = "TooLargeTool") : Logger {

    override fun log(msg: String) {
        Log.println(priority, tag, msg)
    }

    override fun logException(e: Exception) {
        Log.w(tag, e.message, e)
    }

    override fun logTransactionTooLargeDetected(msg: String) {
        Log.e(tag, "TransactionTooLarge detected: $msg")
    }
}
